### PR TITLE
Authenticate on backend

### PR DESCRIPTION
**Pull request by SondreKanstad**
## Background
For our application a standard security layer is crucial to protect the answers to the questions regarding compliance. To do this our plan is to let the users use Oauth 2.0 for authentication and in the long term be able to authorize by setting scopes or roles which is a common feature supported by most Identity providers. This PR will lay the structure for Oauth 2.0 authentication in Ktor and focus on supporting Azure's identity provider service Entra ID. Furthermore, the backend endpoints will be protected, only allowing authenticated users to access them. This PR will solely focus on implementing the bare-minimum for authenticating using Oauth 2.0 and subsequent PR's will improve upon security features and other necessities such as signing out and scopes. 

Oauth 2.0 is complicated and to understand and implement it we have used these resources for learning [Oauth ktor documentation](https://ktor.io/docs/server-oauth.html), [Spring Oauth 2.0 implementation youtube video](https://www.youtube.com/watch?v=us0VjFiHogo), [Oauth 2.0 framework doc](https://datatracker.ietf.org/doc/html/rfc6749) and a couple blogposts. 

For implementation I used a free Identity provider service to test the Oauth 2.0 flow and getting access tokens using my google account [Auth0 Okta service](https://manage.auth0.com/). There are other options such as using [Github OAuth apps](https://github.com/settings/applications/new), or Entra ID or any other identity provider service. 

## Solution
We based the solution on a common Oauth 2.0 flow where the `client` tries to reach a protected part of the `server`, and if the `client` is not authenticated it gets redirected to an `authentication server`. This `authentication server` is an identity provider which can support log-in formats such as a single sign-on *(SSO)*. If the `client` successfully authenticates on the `authentication server` then its redirected back to a callback URL where the `server` gets a *state* and a *code*. The server can then send a request to the `authentication server`'s token endpoint with the *code*, *state*, *client_id* and *client_secret* to generate an *access token*. This token can then be used by the `client` to communicate with the `server` and access resources it is authorized to access. The flow is similar to the diagram below which is called the [Authorization code flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow): 
image

### Create separate PR's for each task and link to this PR in description

- [x] Retrieve cookie / access token and add it as a header (Fixed in [5b86d76](https://github.com/kartverket/regelrett/pull/85/commits/5b86d765448f74e63af61b97f61b088f1b90ab26))
- [x] Automatically redirect unauthenticated users trying to access protected routes to login (https://github.com/kartverket/regelrett/pull/88)
- [x] Change callback URL from localhost to environment variable or a more dynamic URL (fixed in [af05cda](https://github.com/kartverket/regelrett/pull/85/commits/af05cdac306a567066d999046fb6c79f1cb39c37))
- [x] Apply stricter CORS headers (https://github.com/kartverket/regelrett/pull/91)
- [x] Kvibify or remove login page (https://github.com/kartverket/regelrett/pull/88)
- [x] Implement more secure token validation on protected routes (currently only checks if a `user_session` exists (fixed in https://github.com/kartverket/regelrett/pull/114)
- [x] Authorize user with team from token https://github.com/kartverket/regelrett/pull/116
- [x] Check that the implementation works with Entra ID (https://github.com/kartverket/regelrett/pull/90)
- [ ] Be able to develop locally with no authentication and mocked groups (https://github.com/kartverket/regelrett/pull/132)

## How to test

1.  Pull branch
2. Create an Authentication application from an identity provider (For example [Github OAuth apps](https://github.com/settings/applications/new))
3. Set environment variables (An easy approach is setting it in the application configuration in IntelliJ)
4. Build and run both the backend and frontend application
5. Try to access either [http://localhost:3000/team/Team%201](http://localhost:3000/team/Team%201) or [http://localhost:8080/metodeverk](http://localhost:8080/metodeverk)
6. Log in by going to [http://localhost:3000/login](http://localhost:3000/login)
7. Try to access either [http://localhost:3000/team/Team%201](http://localhost:3000/team/Team%201) or [http://localhost:8080/metodeverk](http://localhost:8080/metodeverk) 

### Public environment variables for local testing
image
image

### CODE REVIEW

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            oauth("auth-oauth-azure") {
+                urlProvider = { "http://localhost:8080/callback" }
+                providerLookup = {
+                    OAuthServerSettings.OAuth2ServerSettings(
+                        name = "auth0",
+                        authorizeUrl = System.getenv("AUTH_AUTHORIZE_URL"),
+                        accessTokenUrl = System.getenv("AUTH_ACCESS_TOKEN_URL"),
+                        requestMethod = HttpMethod.Post,
+                        clientId = System.getenv("AUTH_CLIENT_ID"),
+                        clientSecret = System.getenv("AUTH_CLIENT_SECRET"),
```
Comment (SondreKanstad) [inline backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:40]: Alle disse verdiene får man typisk fra Identity provideren når man setter opp en authentication application. Et eksempel kan sees her fra Okta Oauth 
image

image

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import io.ktor.serialization.kotlinx.json.*
+import io.ktor.server.application.*
+import io.ktor.server.auth.*
+import io.ktor.server.sessions.*
+
+val applicationHttpClient = HttpClient(CIO) {
+    install(ContentNegotiation) {
+        json()
+    }
+}
```
Comment (KrisAesoey) [inline backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:16]: I min PR gjorde eg det direkte i modulen. Er det en annen fordel av å gjøre det i klienten?

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+fun Application.installSessions() {
+    install(Sessions) {
+        cookie<UserSession>("user_session") {
+            cookie.path = "/"
+            cookie.maxAgeInSeconds = 360
+            cookie.secure
+            cookie.httpOnly = true
```
Comment (KrisAesoey) [inline backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:24]: Funker bare localhost da? Hva skjer med prod som eg antar vil bruke https 🤔

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt [hunk 4]
```diff
@@ -39,115 +42,175 @@ fun Application.configureRouting() {
...
-            val data = airTableController.fetchDataFromAlle()
-            call.respondText(data.records.toString())
+            val userSession: UserSession? = call.sessions.get<UserSession>()
+            if (userSession != null) {
+                val data = airTableController.fetchDataFromAlle()
+                call.respondText(data.records.toString())
+            }
+            call.respond(HttpStatusCode.Unauthorized)
         }
     }
```
Comment (KrisAesoey) [inline backend/src/main/kotlin/no/bekk/plugins/Routing.kt:68]: Eg antok vi kom til å bruke en slags authentication annotasjon på routingen for å håndtere dette. Det virker som at basert på [denne](https://ktor.io/docs/server-auth.html#authenticate-route) dokumentasjonen at man kan definere providers, som man kan kalle på med en `authentication` som beskriver hvilken provider man ønsker å bruke. Da tror eg man kan bare ta den funksjonaliteten du har laget med å sjekke userSession, men gjøre det til en standardisert måte å autentisere, som kan lett gjenbrukes og se at blir brukt på alle endepunktene 😃

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt [hunk 5]
```diff
@@ -39,115 +42,175 @@ fun Application.configureRouting() {
...
+                databaseRepository.getAnswerFromDatabase(answer)
+                call.respondText("Answer was successfully submitted.")
+            }
+
+            call.respond(HttpStatusCode.Unauthorized)
+        }
+    }
+
+    routing {
+        authenticate ( "auth-oauth-azure" ) {
```
Comment (KrisAesoey) [inline backend/src/main/kotlin/no/bekk/plugins/Routing.kt:151]: Vent hæ, no forvirrer eg meg selv her. Ref kommentaren over med å bruke providers. Du har jo satt det opp her, men kan man ikke gjøre det for alle endepunkt da? Eller er det noe som gjør at det bare vil funke på login endepunktene?

File: frontend/beCompliant/src/components/Login.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  };
+
+  useEffect(() => {
+    window.location.href = "http://localhost:8080/login";
+  }, []);
+
+  return (
+    <div>
+      <p>Redirecting to login...</p>
+    </div>
```
Comment (KrisAesoey) [inline frontend/beCompliant/src/components/Login.tsx:21]: Kvib-ify denne kanskje?

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt (hunk 2 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:16]: Ikke som jeg vet, greit at det er likt tenker jeg

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt (hunk 3 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:24]: Tror http only bare er med på å beskytte tokens som serveres over [Owasp http only](http https://owasp.org/www-community/HttpOnly)

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt (hunk 4 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/plugins/Routing.kt:68]: Ja enig, er ganske sikker på at det er et sånn patern man skal bruke

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt (hunk 5 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/plugins/Routing.kt:151]: Litt usikker, fordi i tutorialen så har man kun `login` og `callback` endepunktene som er innenfor den provideren, også var det protected endepunktet utenfor provideren, men sjekker kun `user_session`. [Ktor eksempel](https://github.com/ktorio/ktor-documentation/blob/2.3.12/codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt). Kan også være at eksemepelet ikke er best practice, og at det kun er et enkelt eksempel, men jeg tror da endepunktene må ligge i en annen provider

File: frontend/beCompliant/src/components/Login.tsx (hunk 6 above)
Comment (SondreKanstad) [reply frontend/beCompliant/src/components/Login.tsx:21]: Lurer på om vi ikke trenger filen i det hele tatt, siden det er en automatisk redirect 🤔

File: backend/src/main/kotlin/no/bekk/Application.kt [hunk 7]
```diff
@@ -22,7 +23,9 @@ fun main(args: Array<String>) {
 }
 
 fun Application.module() {
-    configureRouting()
     configureCors()
     runFlywayMigration()
-}
+    installSessions()
+    initializeAuthentication()
+    configureRouting()
```
Comment (perzonas) [inline backend/src/main/kotlin/no/bekk/Application.kt:30]: Does routing need to be configured after authentication or why was it moved? 😄

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        }
+    }
+}
+
+fun Application.initializeAuthentication(httpClient: HttpClient = applicationHttpClient) {
+    install(Authentication) {
+            oauth("auth-oauth-azure") {
+                urlProvider = { "http://localhost:8080/callback" }
```
Comment (perzonas) [inline backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:32]: Should this be an environment variable as well?

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            cookie.maxAgeInSeconds = 360
+            cookie.secure
+            cookie.httpOnly = true
+        }
+    }
+}
+
+fun Application.initializeAuthentication(httpClient: HttpClient = applicationHttpClient) {
+    install(Authentication) {
+            oauth("auth-oauth-azure") {
```
Comment (perzonas) [inline backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:31]: Should this be an environment variable?

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt (hunk 4 above)
Comment (perzonas) [reply backend/src/main/kotlin/no/bekk/plugins/Routing.kt:68]: Ja min første tanke var at dette ikke ser riktig ut uten at ha jobbet med noe av dette tidligere. Men noe som er enklere en at repetere dette på hvert endepunkt i hele løsningen føles feil hvertfall

File: backend/src/main/kotlin/no/bekk/plugins/Routing.kt (hunk 5 above)
Comment (perzonas) [reply backend/src/main/kotlin/no/bekk/plugins/Routing.kt:151]: kan man wrappe parent routen i authenticate og alle child routes blir da protected?

Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/plugins/Routing.kt:151]: Ja tipper det er det som er standard egentlig, typ en annen authenticate("protected-routes") hvor man setter opp en custom sjekk på access token osv

File: backend/src/main/kotlin/no/bekk/Application.kt (hunk 7 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/Application.kt:30]: I don't know if its necessary, but Routing uses functions from Authentication. So for me it just made sense to place them "secquentially"

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt (hunk 9 above)
Comment (SondreKanstad) [reply backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:31]: Which part?

**Review by perzonas — APPROVED** [review]

File: backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt (hunk 9 above)
Comment (perzonas) [reply backend/src/main/kotlin/no/bekk/Authentication/Authentication.kt:31]: "auth-oauth-azure"

**Pull request merged**